### PR TITLE
Add custom properties to CZML files

### LINKED ISSUE #3162: Support for custom CZML properties
We want to add custom properties to CZML files. Currently we are doing this with a workaround, by wedging some json into the `description` field. @kring and I spent some time today discussing alternatives, and would like to get your feedback on the scheme we came up with. If you are ok with it, one of us can then submit a pull request.

A few explanatory points -
- We decided against adding custom properties at the root level of the CZML object, since they might conflict with current or future special properties (eg. `point`, `polygon`). So instead we propose they appear under a new `properties` key.
- We want to maintain the basic idea that CZML properties can be time-varying. So that the custom properties do not have to all vary in the same way, we will not define intervals on `properties` directly, but instead on its sub-properties.
- Those sub-properties would be interpreted using a new "generic" type handler (in [`unwrapInterval`](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/DataSources/CzmlDataSource.js#L342-L382)). Properties whose values are arrays or objects would need to put that value on the special key `value`. (Much like the current color handler requires a special key `rgba`.)
- For simplicity, nested properties cannot have different intervals to their parent properties, and so do not need the extra `value` key. This could be allowed in an extension later.

Here's an example:

```
"properties": {
    "short_name": "ABC",
    "changing_name": [
        {
            "interval": "2007-03-02T13:00:00Z/2008-05-11T15:30:00Z"
            "value": "XYZ"
        },
        ...
    ],
    "foobar": {
        "value": {
            "foo": 4,
            "bar": 10
        }
    },
    "an_array": {
        "value": [...]
    },
    "children": {
        "value": {
            "name": "alice",
            "children": [
                {
                    "name": "bob",
                    "children": []
                }
            ]
        }
    }
}
```

What do you think?

### PR DESCRIPTION
**Pull request by RacingTadpole**
Fixes #3162.

Here's a summary -
- We decided against adding custom properties at the root level of the CZML object, since they might conflict with current or future special properties (eg. `point`, `polygon`). So instead we propose they appear under a new `properties` key.
- We want to maintain the basic idea that CZML properties can be time-varying. So that the custom properties do not have to all vary in the same way, we will not define intervals on `properties` directly, but instead on its sub-properties.
- Nested properties cannot have different intervals to their parent properties, and so do not need the extra `value` key.

For example:

```
"properties": {
    "short_name": "ABC",
    "changing_name": [
        {
            "interval": "2007-03-02T13:00:00Z/2008-05-11T15:30:00Z"
            "value": "XYZ"
        },
        ...
    ],
    "foobar": {
        "value": {
            "foo": 4,
            "bar": 10
        }
    },
    "an_array": {
        "value": [...]
    },
    "children": {
        "value": {
            "name": "alice",
            "children": [
                {
                    "name": "bob",
                    "children": []
                }
            ]
        }
    }
}
```

I realize this is a change to the CZML spec, so feedback/suggestions appreciated :-) .

### CODE REVIEW

**Comment by mramato** [comment]
I haven't looked at this yet, but I'm sure it will make a lot of people happy.  I'll try to find some time to review this weekend.

@shunter Do you think this approach will jive nicely with the schema and be easy to implement in [czml-writer](https://github.com/AnalyticalGraphicsInc/czml-writer)?  Does the recent PropertyBag stuff we added for models come into play here?

**Comment by pjcozzi** [comment]
This sounds fine to me; in glTF, this would be `extras`, not `properties`, but I'm not sure that CZML and glTF need to be consistent here.

@mramato is it much work on our end to get this merged?

**Comment by mramato** [comment]
Sorry, I just haven't made the time to look at this in more detail; however I'd really like to see this be officially supported so I'll try and get with @shunter in my spare time and provide feedback.

**Comment by RacingTadpole** [comment]
Hi guys - can I keep this one (Adding custom properties to CZML files) on your radar?

**Comment by hpinkos** [comment]
Thanks @RacingTadpole 

@mramato @shunter what's the status on this one?

**Comment by kring** [comment]
@mramato consider reviewing this one for the bug bash?  We've been using it for close to a year now and are quite happy with it.

**Comment by kring** [comment]
@RacingTadpole mind merging in master in case someone does have a chance to review this during the bug bash?

**Comment by mramato** [comment]
> @mramato consider reviewing this one for the bug bash? We've been using it for close to a year now and are quite happy with it.

I'm working my way through old PRs, so I would definitely like to look at this.  I was hoping @shunter would have some thoughts since he has done a lot of CZML work lately, but barring any major issues, I don't see why we couldn't get this in soon.

**Comment by RacingTadpole** [comment]
Great, thanks guys. I've merged in the latest master now.

**Comment by mramato** [comment]
I haven't forgotten about this PR, I spoke a little offline with @shunter and we have a few concerns with how this ends up working with `CompositeEntityCollection` and a few other issues that may make it harder to implement as part of `czml-writer`  I also want to make sure that the Entity API ends up being consistent with custom properties from various datasources and not have `properties` be CZML specific.  I'll definitely spend some time fleshing out these thoughts later this week and maybe even prototype some code.  I'll keep you posted.

**Comment by kring** [comment]
Perhaps you could describe your concerns, and maybe we can help address them?

**Comment by mramato** [comment]
> Perhaps you could describe your concerns, and maybe we can help address them?

Absolutely, sorry didn't mean to indicate I was just going to finish this myself.  I just need to actually spend some time investigating all of the implications since I couldn't like everything off the top of my head.

**Comment by kring** [comment]
Ok cool, thanks for taking a look at it!

**Comment by pjcozzi** [comment]
What is the plan with this pull request?

@kring @RacingTadpole we probably don't have a ton of bandwidth to help you fully flush this out in the next few months.  From your perspective, is this ready to go?  Should work continue in this PR or should this be tabled until @mramato or @shunter has cycles to help?

**Comment by kring** [comment]
@pjcozzi as far as we know this is ready to go.  I worked closely with @RacingTadpole to design it, and I was heavily involved in the original design of CZML and am pretty aware of the issues there, so I don't think we came up with something unworkable or abusive here.  @mramato said he and @shunter had concerns, but never really said what they are.

**Comment by shunter** [comment]
@kring it's been a while since I talked about this with @mramato (sorry!) but it was less about concerns and more about making sure that details like property change notification (which is relied on by the composite entity collection) are hooked up.  The `PropertyBag` class may help out here rather than just using an object literal.  I don't know if the model node transformations feature was in CZML at the time, but there are similarities in the features, so keeping consistency would be good.

Also I note that the custom properties concept doesn't exist at the Entity level.  I imagine other types of data sources may have information which would naturally want to use the same approach (though I suspect most other data sources wouldn't have time-dynamic custom data, but I don't know)

Beyond that would be making sure that this approach will fit well into the czml-writer architecture (such as it is).  Once it's in Cesium, it's de facto part of the CZML "spec" (such as it is) making it problematic to change, so I'd want to make sure that these properties will be easy to work with, both in the czml-writer, and in STK Components.

I had hoped @mramato would have had time to look into this, but if not I can try to carve out some time when I can.  Like I said, my main hesitance is simply because once it's in the "spec" it's permanent.

**Comment by kring** [comment]
Thanks @shunter!  I've just pushed changes to add `properties` to `Entity` and to add a test to ensure that it works with `CompositeEntityCollection`.  As expected, I had to make some changes for notifications to work right.

A few non-obvious consequences:
1. This is a breaking change for anyone using `GeoJsonDataSource` and accessing the `properties`.  Previously every property in there was just a regular raw value.  Now it's a `ConstantProperty`.
2. I had to change `PropertyBag` to handle the possibility of a `null` value, which is valid and different from undefined.

re: 1, the only ways I can think of to avoid this are either:
* Don't automatically add `properties` to `Entity`.  Let `CzmlDataSource` or whoever else wants it on their entities add it themselves (this is how `GeoJsonDataSource` worked before this PR).
* Name `properties` something else, like `extras`.  But then `GeoJsonDataSource` will have both `properties` and `extras`, which is weird.

re: czml-writer, I don't really know what's involved (I don't even have Visual Studio these days) but hopefully now that `properties` is using `PropertyBag`, it will work in czml-writer the same way node transformations work?

**Comment by kring** [comment]
In addition to `PropertyBag`, I had to change `createPropertyDescriptor` to allow null as a valid property value distinct from undefined.  Without that change, a `GeoJsonDataSource` with a value of `null` for one of its `properties` would cause the raw value `null` to appear in the `Entity`'s property bag, instead of a `ConstantProperty` with the value null, which was weirdly inconsistent (and caused a test failure in `GeoJsonDataSourceSpec`).

**Comment by shunter** [comment]
re: GeoJson properties being `ConstantProperty` I do think your change is consistent.  Other possible ways to mitigate the breaking change:

1. We could add `valueOf` to `ConstantProperty` to coerce to primitive by unwrapping them.  That would probably need some experimentation to see what unexpected results pop up, though.  `valueOf` is not quite as good as an implicit conversion operator in real languages.
2. `PropertyBag` itself has a `getValue` which will effectively unwrap all the contained properties.  Rather than calling `getValue` on each custom property in our example code and sandcastle, etc., it might be easier to justify by suggesting that when using custom properties (especially in GeoJson), use `getValue` on the whole bag.

**Comment by shunter** [comment]
I notice that you don't currently have any tests that use sampled custom properties.  I'm assuming that nothing will prevent it, but it would be good to have demonstrated.

I'm going to dig into trying to figure out how this will work in the czml-writer today.  Hopefully I can get some additional sample documents which we can put into tests.

**Comment by shunter** [comment]
OK, after working on this a bit from the writer side, it looks like sampled properties, or any type of typed properties, aren't hooked up yet.

For reference, here's a sample document from what I have so far:
```json
[
  {
    "id":"document",
    "version":"1.0",
    "clock":{
      "interval":"2012-04-02T12:00:00Z/2012-04-02T12:01:00Z",
      "currentTime":"2012-04-02T12:00:00Z"
    }
  },
  {
    "id":"MyID",
    "properties":{
      "custom_boolean":[
        {
          "interval":"2012-04-02T12:00:00Z/2012-04-02T12:00:01Z",
          "boolean":true
        },
        {
          "interval":"2012-04-02T12:00:01Z/2012-04-02T12:00:02Z",
          "boolean":false
        },
        {
          "interval":"2012-04-02T12:00:02Z/2012-04-02T12:01:00Z",
          "boolean":true
        }
      ],
      "custom_cartesian":{
        "epoch":"2012-04-02T12:00:00Z",
        "cartesian":[
          0,1,2,3,
          60,4,5,6,
          120,7,8,9
        ]
      }
    }
  }
]
```

**Comment by kring** [comment]
> OK, after working on this a bit from the writer side, it looks like sampled properties, or any type of typed properties, aren't hooked up yet.

Yeah, that's mostly by design.  The way CZML is parsed right now, the type of a property must be known ahead of time.  For example, `CzmlDataSource` knows that `polygon.outlineColor` is a color and parses it accordingly.  Never do we "look inside" an interval to deduce its type.

But in the case of properties, we don't know what type to expect for any given property.  So our approach here was to say that the type of a property inside `properties` is a special type where its value is given with the `value` property.  The value can be any valid JSON type.  `value` can never be interpolated.  We wouldn't be able to tell time-tagged values from a single value which is an array.

In theory we can extend this special type to support other keys than `value` with different meaning.  e.g. `number` could work the way you've used it above.  This will require us to peer into intervals (the first one?) to deduce the type, but it might be ok.  I don't think it's trivial to work this into `CzmlDataSource`, though, and I also see a clear compatibility path for adding that in the future, so I don't think it's important to add now.

**Comment by kring** [comment]
By the way, while sampling of custom properties isn't supported, intervals are.  You just need to change `"boolean": true` to `"value": true` in your example.

**Comment by mramato** [comment]
Sorry for being MIA lately.  I pretty much agree with @shunter on everything (no surprises there 😄) I think it's important for custom properties to support all official property types.  I honestly think this will be fairly trivial to add from a client code standpoint, so I can explore this sometime soon (possible during the bug bash if not this weekend).  I'll branch of this PR with some ideas and let you guys know when I have something.

**Comment by kring** [comment]
Sure, I don't think I disagree with much Scott said either!

I'm remembering now that @RacingTadpole and I did spend awhile talking about this problem when he was first working on it a year ago.  There are basically two ways we could have stronger typing for custom properties (which allows sampled properties to work and maybe has other advantages):

1. Make the client "peek inside"

**Comment by shunter** [comment]
My main motivation here is consistency, and not making custom properties special or different.  I would expect a custom boolean property to be represented in exactly the same way as any other boolean property that already exists, like `show`.  The approach of specifying type information in an entirely new way goes against that.  Similarly, I would expect custom properties to have all the same capabilities as predefined properties, and vice versa.

I'm not clear on the objection to pulling the type information out of the document itself (looking inside the interval), given that the current implementation largely already works this way.  You're right that it knows a priori that `position` is a `Cartesian` property, but recall that `Cartesian` properties can actually be specified in a number of ways: `cartesian`, `cartographicRadians`, `cartographicDegrees`, `cartesianVelocity`, `reference`.  When we're parsing, we look in the interval to see which of these properties is actually present in the document and parse/convert accordingly.

Basically, the current code has a mapping from JS type to a value property name, or names.  Making this mapping bidirectional seems like it should be straightforward, and I think will produce a better result.

**Comment by pjcozzi** [comment]
Thanks for kicking this off!  This is now included in #5105.
